### ISSUE TITLE
`exec` source: redirect children stderr to parent stderr when `include_stderr` is `false`

### ISSUE THREAD

**Issue by hhromic** [vectordotdev/vector#12368]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

I'm sure I have mentioned this feature request before and somewhere but I couldn't find any open issue. Feel free to close this issue if you manage to find a duplicate.

The use-case for this feature request is to improve the observability of the processes started by the `exec` source when `include_stderr` is set to `false`, i.e. when we are not interested to process data from `stderr`.

Most of the time, `stderr` is used for diagnostic output from most processes, and it would be very handy, convenient and succint for Vector to simply relay any children output on `stderr` to its own `stderr` when the user is not interested on processing data from `stderr`.

For example, consider the following topology:
```yaml
sources:
  exec:
    type: exec
    command: ["bash", "-c", "echo 'oops, something is wrong' > /dev/stderr; echo 'all is good' > /dev/stdout"]
    mode: scheduled
    include_stderr: false
    scheduled:
      exec_interval_secs: 1
sinks:
  console:
    type: console
    inputs: ["exec"]
    encoding: json
```

Of course, I'm just using a toy `command` here but a common real-world use-case is calling `curl`.

In this topology, we are only interested on processing actual output from `stdout` and leave diagnostic messages from the spawned process to go to the error facility, i.e. of Vector. However, today this is not possible because when `include_stderr` is `false`, Vector simply ignores any output from `stderr` silently.

This makes any error coming from the child processes to be invisible and hard to monitor/observe.
See attempted solutions.

### Attempted Solutions

There are two possible solutions currently in Vector, none of them 100% satisfactory.
Both of them involve capturing `stderr` with `include_stderr` set to `true` but differ in how the output stream is handled.

### Do not do anything special about data in `stderr`

The first option is to simply look aside and do nothing special about potential errors in the stream.
In this case, we can let any transform to just error out with "invalid data" from `stderr`. This is especially "robust" when the actual data from `stdout` is in a well-known format (e.g. JSON) and the data in `stderr` is not.
```yaml
transforms:
  parse-json:
    type: remap
    inputs: ["exec"]
    source: |
      ., err = parse_json(.message)
      if err != null {
        log("invalid JSON message: " + string!(.message), level: "warn")
        abort
      }
```
OR
```yaml
sources:
  exec:
    type: exec
    include_stderr: true
    decoding:
      codec: json
```

These will error out for any diagnostic message that is not JSON, so we will know _something_ is wrong.

However, the actual diagnostic message is hard to read. Especially the second example where the source is configured to parse JSON itself. The source will not print out the offending message. The first example (using `remap`) will reveal the diagnostic error in a confusing manner, i.e. a JSON parsing error.

### Separate the `stdout` and `stderr` streams and route to `console`

To improve on the previous approach, we can instead `route` the streams and send `stderr` to a console sink using `stderr`:
```yaml
sources:
  exec:
    type: exec
    command: ["bash", "-c", "echo 'oops, something is wrong' > /dev/stderr; echo 'all is good' > /dev/stdout"]
    mode: scheduled
    include_stderr: true
    scheduled:
      exec_interval_secs: 1

transforms:
  route:
    type: route
    inputs: ["exec"]
    route:
      stdout:
        type: check_fields
        stream.eq: stdout
      stderr:
        type: check_fields
        stream.eq: stderr

sinks:
  console:
    type: console
    inputs: ["route.stdout"]
    encoding: json
  stderr:
    type: console
    inputs: ["route.stderr"]
    target: stderr
    encoding: text
```

This second approach is much nicer and somewhat elegant in the sense that it will transparently relay any diagnostics output from the spawned process in `stderr` to Vector's own `stderr`. Thus any errors/diagnostics from the children processes are visible exactly how they were meant.

This is quite equivalent to doing shell piping in Unix:
```
cat /path/to/file | grep pattern | awk 'script' > /path/to/output
```
Because the `stderr` of all these programs is connected to terminal transparently, you will see any errors/diagnostics from them without doing anything special.

I believe that Vector should follow the same approach when `include_stderr` is `false`

### Proposal

My proposal is to implement the second approach described above natively in Vector to avoid all the routing/sinking boiler plate :)

In other words, for Vector to transparently connect its `stderr` with the children's `stderr` when `include_stderr` is `false`.

A more advanced implementation could relay `stderr` from children processes to Vector's own internal logging system instead, so it can be processed using the `internal_logs` source. I'm not particularly interested on this capability but other users might find it useful.

### Version

vector 0.21.0 (x86_64-unknown-linux-gnu c1edb89 2022-04-14)

**Comment by jszwedko** [comment]
Thanks for this write-up @hhromic !

As a work-around, the `data_stream` is added as metadata to each log event and so the `route` transform can be used after the `source` in a manner similar to your separate output streams for `exec`.

I am kinda partial to your suggestion of a separate output to let users route it how they see fit. While I can see the use of passing through `stderr` when it is not being captured, I could see this being confusing to users that expect Vector's output to only be its own logs (the format of the subprocess logs is also unlikely to match Vector's own). It also opens up additional questions like the one you raised about whether they should appear `internal_logs`.

We could always follow-up later to allow pass-through in addition to the separate output stream approach identified above, but I think my preference would be to start with that.

**Comment by hhromic** [comment]
> Thanks for this write-up @hhromic !
> 
> As a work-around, the `data_stream` is added as metadata to each log event and so the `route` transform can be used after the `source` in a manner similar to your separate output streams for `exec`.
> 

Hey @jszwedko, thanks for replying!
I have to admit that I don't follow what you meant here :(
Can you elaborate a bit more the workaround you suggest so I can understand it?
It sounds like the second "attempted solution" I described but using metadata instead of the `stream` field?
If that is the case, how that would be different than using the `stream` field and the `route` transform I described?

> I am kinda partial to your suggestion of a separate output to let users route it how they see fit. While I can see the use of passing through `stderr` when it is not being captured, I could see this being confusing to users that expect Vector's output to only be its own logs (the format of the subprocess logs is also unlikely to match Vector's own). It also opens up additional questions like the one you raised about whether they should appear `internal_logs`.
> 

Yes, I totally understand your position there (and also agree!) and why I also mentioned a "more advanced implementation" where lines received from `stderr` (when `include_stderr` is `false`) could be instead redirected through Vector's native logging system instead of naively redirecting to Vector's own `stderr` (and thus logging getting mixed).

One of the headaches I can see is, for example, what logging level to use for these messages. Info? Warn? Error? Despite the name `stderr`, not everything coming in this stream is necessarily an error, and hence why the term "diagnostics" is more appropriate :)

> We could always follow-up later to allow pass-through in addition to the separate output stream approach identified above, but I think my preference would be to start with that.

If a pass-through via the Vector logging system is implemented, I wouldn't feel the need for a "plain" pass-through.

**Comment by jszwedko** [comment]
Hey!

> I have to admit that I don't follow what you meant here :(
> Can you elaborate a bit more the workaround you suggest so I can understand it?
> It sounds like the second "attempted solution" I described but using metadata instead of the `stream` field?
> If that is the case, how that would be different than using the `stream` field and the `route` transform I described?

Doh, you are right, I misread that section. I mean the same thing as you have under "Separate the stdout and stderr streams and route to console".

My suggestion was that we could also enhance the `exec` source to have the streams separated already (similar to recent changes to the `remap` transform) so that you could just have:

```yaml
sources:
  exec:
    type: exec
    command: ["bash", "-c", "echo 'oops, something is wrong' > /dev/stderr; echo 'all is good' > /dev/stdout"]
    mode: scheduled
    include_stderr: true
    scheduled:
      exec_interval_secs: 1

sinks:
  console:
    type: console
    inputs: ["exec.stdout"]
    encoding: json
  stderr:
    type: console
    inputs: ["exec.stderr"]
    target: stderr
    encoding: text
```

Alleviating the need for `route`. Granted, it is a small UX improvement in this case since `route` allows it to be expressed pretty simply.

**Comment by hhromic** [comment]
> Doh, you are right, I misread that section. I mean the same thing as you have under "Separate the stdout and stderr streams and route to console".

Ah I see, yes that was the most natural way I could think of for implementing this with current Vector. 👍

> My suggestion was that we could also enhance the `exec` source to have the streams separated already (similar to recent changes to the `remap` transform) so that you could just have:

Ah I think you mean adopting _named outputs_ for the source? If so, I think is quite a good idea!
In other words, you suggest that the `exec` source can have the following named outputs?
* `_default`: outputs stdout (and stderr if `include_stderr` is `true`) from the exec'd process, i.e. same as it is today.
* `stdout`: only outputs stdout from the exec'd process (`include_stderr` would be irrelevant here)
* `stderr`: only outputs stderr from the exec'd process (`include_stderr` would be irrelevant here)

That would make the source quite elegant indeed and clear to route around the topology. 💯 

The user could then (if needed) route to a `remap` using `log()` with any desired log-level in case logging to Vector's logging system is desired, or simply route to a `stderr` blackhole like in your example. Very little "boiler-plate" indeed.

That sounds quite nice to me!

**Comment by hhromic** [comment]
I was just going over the existing source components and their outputs, and I think the `datadog_agent` source is a very nice example of what I think you propose: https://vector.dev/docs/reference/configuration/sources/datadog_agent/#outputs

So in summary, we can add a `multiple_outputs` config to `exec` (mirroring how is done in `datadog_agent`), and:
* When set to `true`, the source uses the `stderr` and `stdout` outputs (and not `<component_id>`).
* When set to `false` (the default so is opt-in) it just outputs to `<component_id>` as it is today.

In other words, fully mirror the approach in `datadog_agent`.

If you think that design is a good idea, I think I can work on that myself by mirroring how is implemented in `datadog_agent`.
